### DISCUSSION TITLE
Malicious Counter errors on apple devices only

### DISCUSSION THREAD

**Discussion by davidbairdala** [webauthn4j/webauthn4j#782] [Q&A] [answered]
Hi,

I feel like I must be missing something obvious here, but I can't figure it out. I am using webauthn4j to provide device authentication for my web app (grails). I store the server-side authenticator data in a database and use browser local storage to store a credential id in the client so that when users come back (after registering their device) they can be presented with a "log in using this device" button which kicks off the webauthn authentication flow.

When I use my Android phone or tablet, or my Windows PC this process works flawlessly. However, when I test using an Apple device (devices tested include iPhone 14, iPhone 14 pro and a 2019 macbook pro), users are able to register the device and log in once using the device (face id or passcode etc). The second time they try and log in, however, webauthn4j throws an error "Malicious counter value is detected. Cloned authenticators exist in parallel.". Through some logging, I can confirm that the server-side counter is 1, but the passed counter from the client is always 0. When I check the logs after using my Android devices I see that both counters are being incremented, and the server-side counter is always one less than the client counter, as expected. So I don't think my code is fundamentally wrong, I just think I have missed some detail that is causing the issue on Apple devices.

I am using the non-strict authentication manager - could this be the reason? I basically structured my code off the scaffolding in the quick start guide (https://webauthn4j.github.io/webauthn4j/en/#quick-start), so I believe it is basically correct. Has anyone else had this kind of problem with Apple devices specifically?

Cheers, David

**Comment by davidbairdala** [comment]
Ok, So I think I have an answer to my own question. Starting with iOS 16 and presumably the most recent version of OS X, Apple have implemented webauthn using their "passkeys". More info here: https://www.slashid.dev/blog/passkeys-deepdive/

The whole point of Apple passkeys are that they are synchronized across devices, so according this thread (https://developer.apple.com/forums/thread/716954) the signing count kind of defeats that purpose, and so they will always return 0 for that.

So I suppose I have the option of replacing the Default malicious signing count behaviour with one that doesn't actually care that the counts are different. Ideally, I'd like to disable this check just for Apple devices though, but I'm not sure if there is a reliable way to actually determine the device type, other than maybe the User-Agent?

Any advice would be greatly appreciated.

**Reply by ynojima** [reply, marked as answer]
Hello, thank you for reporting the issue.

How do you update the `counter` value of `Authenticator` after the authentication (assertion verification)?
If you increment the `counter` value, it is the cause. You need to update the `counter` value with `authenticationData.getAuthenticatorData().getSignCount()` value.
Please see the last line of the sample code: https://webauthn4j.github.io/webauthn4j/en/#webauthn-assertion-verification

As you have already checked, Apple's device always returns 0 from `authenticationData.getAuthenticatorData().getSignCount()`, the `counter` value of `Authenticator` is always kept to 0, and the issue will not happen.

**Reply by davidbairdala** [reply]
Thanks so much for your reply. You are right, I was simply incrementing the counter, not storing the sign count as returned by the authenticator data.
